### PR TITLE
Build: Exit when wrong node version is active

### PR DESCRIPTION
**Pull request by akirk**
This is to ensure that a new developer doesn't run into problems when using a non-supported node version. Instead of continuing with the process, it will now bail out like this:

```
> wp-calypso@0.17.0 prestart /Users/alex/Sites/Automattic/wp-calypso
> npm run -s install-if-deps-outdated && node bin/welcome.js && npm run -s check-versions

             _                           
    ___ __ _| |_   _ _ __  ___  ___      
   / __/ _` | | | | | '_ \/ __|/ _ \ 
  | (_| (_| | | |_| | |_) \__ \ (_) |  
   \___\__,_|_|\__, | .__/|___/\___/ 
               |___/|_|                

node: 8.5.0
npm: 5.3.0
yarn: 1.0.2
Error: Wanted node version 6.11.2 (6.11.2)
To install node, run `nvm install 6.11.2` or see https://nodejs.org/
Error: Wanted npm version 3.10.10 (3.10.10)
To install npm, run `npm install -g npm@3.10.10`

!!! Calypso will only work with the versions listed. Please set up your environment accordingly !!!
```

Maybe there are good reasons not to go this route, this is mainly to remove possible onboarding frictions.

### CODE REVIEW

**Comment by gwwar** [comment]
Did folks test with wp-desktop? They're actually slightly ahead of us due to electron. cc @astralbodies @spen

**Comment by spen** [comment]
@gwwar I'm fairly sure that we'll be ok here since the node version that ships with the app is the one we had to update. The node version that we use for building can happily be a different version to the one Electron uses.

I'll double check this when I'm back from the ⛰s though :)

**Comment by akirk** [comment]
@spen Do you have an update for me here? Thanks!

**Comment by spen** [comment]
Hey @akirk! Following a conversation with @aidvu I started working on https://github.com/Automattic/wp-desktop/pull/338 which aims to encourage parity of the versions of NPM and Node between the two projects.
My earlier point about the versions (23 Sept) Is only half true. The version that we ship with the desktop app (The version that Electron determines) Can be different to either of those that `wp-desktop` or `wp-calypso` state in their `package.json`'s without consequence.  
However, given that part ofthe build process of `wp-desktop` involves `cd`ing in to calypso and triggering the calypso build - we risk doing doing so with an unexpected working version (unexpected by Calypso) because we're currently working with the version that `wp-desktop` demands.

One solution would be to maintain parity with calypso at all time another might be to somehow switch the working node version part-way through the build (I have no idea how this would work though)

**Comment by DanReyLop** [comment]
> Maybe there are good reasons not to go this route, this is mainly to remove possible onboarding frictions.

I remember having to explicitly add the `exit 0` so the build would keep working, by popular demand. The reason is that we have a handful of NodeJS-based projects, and they aren't perfectly in sync, version-wise. I'm a few patch versions behind and that's ok, Calypso forbidding me to build if I'm in `6.11.1` because it's requiring `6.11.2` can be really annoying.

There are tools that make version-switching less annoying (`nvm`, `n`), but for example there are no good options for Windows.

Does `check-node-version` accept a `semver` range? Requiring `6.11.*` should be reasonable, or `6.11+` or something like that. It's extremely unlikely that a semver-patch would make any difference.

**Comment by gwwar** [comment]
> I remember having to explicitly add the exit 0 so the build would keep working

In the first iteration I remember this being requested as well. Would it help if we made the warning more noticeable, ie with ascii art or something?

**Comment by spen** [comment]
I'd argue that ascii art might be missed - I know that often when I'm running a build I'll trigger it and do something else while the process takes place since I know it's a slow one. How about requiring confirmation? Something like:

```
You're using x version instead of the recommended y version. Building Calypso with an unexpected version could lead to unexpected problems...

Are you sure you'd like to continue using this version?
```

It'd be pretty awesome if at this point we could offer a choice like `- switch to y version before continuing` that actually do just that 💭

**Comment by apeatling** [comment]
@akirk Is this one actually ready to merge?

**Comment by akirk** [comment]
@apeatling I don't see a consensus on whether it's ok to merge. I'd rather close it for now.

**Comment by dmsnell** [comment]
I'd like to argue here _for_ merging. As annoying as it can be to have to switch to the chosen `node` version, it's even worse when we fail to notice problem on account of having used a different language than what what the project is using.

File: package.json [hunk 1]
```diff
@@ -221,7 +221,8 @@
     "rm": "node bin/rm.js",
     "preshrinkwrap": "npm prune",
     "postshrinkwrap": "node bin/trim-shrinkwrap.js && touch node_modules",
-    "prestart": "npm run -s install-if-deps-outdated && node bin/welcome.js && ( check-node-version --package || exit 0 )",
+    "check-versions": "check-node-version --package || ( echo '\n!!! Calypso will only work with the versions listed. Please set up your environment accordingly.\n' && exit 1 )",
```
Comment (DanReyLop) [inline package.json]: *If* we decide to merge this, `"check-node-version --package"` will just fail with an error, no need to do all that `|| echo && exit 1` contraption.

Comment (akirk) [reply package.json]: Yes but part of the initiative for this PR was to make it easier for newbies and provide a decent error message.

**Pull request closed without merging**